source/common/common/envoy_defaults.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+
+#include "source/common/protobuf/protobuf.h"
+#include "source/common/singleton/threadsafe_singleton.h"
+
+#include "tools/protodoc/manifest.pb.h"
+
+// These macros return the value associated with `field_name` in the protobuf `message` if it
+// exists. Otherwise, they return the value associated with `field_name` in the defaults profile if
+// it exists. Otherwise, `default_value` is returned.
+
+#define PROTOBUF_GET_NUMBER_OR_PROFILE_DEFAULT(message, config, field_name, default_value)         \
+  ([&](const auto& msg) -> double {                                                                \
+    if (msg.has_##field_name()) {                                                                  \
+      return msg.field_name().value();                                                             \
+    }                                                                                              \
+    return DefaultsProfile::get().getNumber(config, #field_name, default_value);                   \
+  }((message)))
+
+#define PROTOBUF_GET_MS_OR_PROFILE_DEFAULT(message, config, field_name, default_value)             \
+  ([&](const auto& msg) -> std::chrono::milliseconds {                                             \
+    if (msg.has_##field_name()) {                                                                  \
+      return std::chrono::milliseconds(DurationUtil::durationToMilliseconds(msg.field_name()));    \
+    }                                                                                              \
+    return DefaultsProfile::get().getMs(config, #field_name, default_value);                       \
+  }((message)))
+
+#define PROTOBUF_GET_BOOL_OR_PROFILE_DEFAULT(message, config, field_name, default_value)           \
+  ([&](const auto& msg) -> bool {                                                                  \
+    if (msg.has_##field_name()) {                                                                  \
+      return msg.field_name().value();                                                             \
+    }                                                                                              \
+    return DefaultsProfile::get().getBool(config, #field_name, default_value);                     \
+  }((message)))
+
+namespace Envoy {
+
+struct DefaultsProfile {
+  enum Profile {
+    Performant,
+    Safe,
+  };
+
+  class ConfigContext {
```

Comment:
Please add comments describing the class purpose

Reply:
I've added an overall comment, but this could probably use some additional instructions. I'll work on it